src/awkward/_nplikes/typetracer.py

```diff
@@ -869,6 +869,21 @@ def div_shape_item(self, x1: ShapeItem, x2: ShapeItem) -> ShapeItem:
         assert result * x2 == x1
         return result
 
+    def shape_equals(
+        self, shape1: tuple[ShapeItem, ...], shape2: tuple[ShapeItem, ...]
+    ) -> bool | None:
+        if len(shape1) != len(shape2):
+            return False
+
+        result_unknown = False
+        for x, y in zip(shape1, shape2):
+            if x is None or y is None:
+                result_unknown = True
+            elif x != y:
+                return False
+
+        return None if result_unknown else True
```

Comment:
Only `TypeTracer` actually returns `None`; when it encounters any unknown components the result is either `False` or `None` depending upon the remaining components.